src/hotspot/share/runtime/synchronizer.hpp

```diff
@@ -155,42 +160,21 @@ class ObjectSynchronizer : AllStatic {
 
   // debugging
   static void audit_and_print_stats(bool on_exit);
-  static void chk_free_entry(JavaThread* jt, ObjectMonitor* n,
-                             outputStream * out, int *error_cnt_p);
-  static void chk_global_free_list_and_count(outputStream * out,
-                                             int *error_cnt_p);
-  static void chk_global_wait_list_and_count(outputStream * out,
-                                             int *error_cnt_p);
-  static void chk_global_in_use_list_and_count(outputStream * out,
-                                               int *error_cnt_p);
-  static void chk_in_use_entry(JavaThread* jt, ObjectMonitor* n,
-                               outputStream * out, int *error_cnt_p);
-  static void chk_per_thread_in_use_list_and_count(JavaThread *jt,
-                                                   outputStream * out,
-                                                   int *error_cnt_p);
-  static void chk_per_thread_free_list_and_count(JavaThread *jt,
-                                                 outputStream * out,
-                                                 int *error_cnt_p);
+  static void chk_in_use_list(outputStream* out, int* error_cnt_p);
+  static void chk_in_use_entry(ObjectMonitor* n, outputStream* out,
+                               int* error_cnt_p);
   static void do_final_audit_and_print_stats();
-  static void log_in_use_monitor_details(outputStream * out);
-  static int  log_monitor_list_counts(outputStream * out);
-  static int  verify_objmon_isinpool(ObjectMonitor *addr) PRODUCT_RETURN0;
-
-  static void do_safepoint_work();
+  static void log_in_use_monitor_details(outputStream* out);
 
  private:
   friend class SynchronizerTest;
 
-  enum { _BLOCKSIZE = 128 };
-  // global list of blocks of monitors
-  static PaddedObjectMonitor* g_block_list;
+  static MonitorList   _in_use_list;
+  static jint          _in_use_list_ceiling;
```

Comment:
Can you add some commentary on what this ceiling is as I could not understand its role just by looking at the code. Thanks.

Reply:
How about this:

Reply:
Thanks for the comment. So instead of checking the threshhold on each OM allocation we use this averaging technique to estimate the number of monitors in use? Can you explain how this came about rather than the simple/obvious check at allocation time. Thanks.

Reply:
I'm not sure I understand your question, but let me that a stab at it anyway...

We used to compare the sum of the in-use counts from all the in-use lists
with the total population of ObjectMonitors. If that ratio was higher than
MonitorUsedDeflationThreshold, then we would do an async deflation cycle.
Since we got rid of TSM, we no longer had a population of already allocated
ObjectMonitors, we had a max value instead. However, when the VMs use
of ObjectMonitors is first spinning up, the max value is typically very close
to the in-use count so we would always be asking for an async-deflation
during that spinning up phase.

I created the idea of a ceiling value that is tied to thread count and the
AvgMonitorsPerThreadEstimate to replace the population value that we
used to have. By comparing the in-use count against the ceiling value, we
no longer exceed the MonitorUsedDeflationThreshold when the VMs use
of ObjectMonitors is first spinning up so we no longer do async deflations
continuously during that phase. If the max value exceeds the ceiling value,
then we're using a LOT of ObjectMonitors and, in that case, we compare
the in-use count against the max to determine if we're exceeding the
MonitorUsedDeflationThreshold.

Does this help?